### PR TITLE
Use new basic block source line data

### PR DESCRIPTION
**Pull request by markrtuttle**
The version of cbmc at the tip of develop now includes in its coverage output an xml/json element for each basic block that gives the set of source lines contributing to that basic block.  Prior versions of cbmc encoded this set of source lines in a text string used as the description of the coverage goal corresponding to this basic block.  Parsing this string encoding source lines was too brittle to handle function names used by kani.

This pull request modifies viewer to use the new xml/json elements if they exist, and to revert to paring the text string if they do not.    It also advances version 3.0.1 -> 3.1.

Testing: For cbmc version 5.56.0 and cbmc version at the tip of develop, I have compared the output of viewer 3.0.1 and 3.1 on the coreHTTP proofs and found the outputs to be identical.  So the new 3.1 works with the xml/json elements (produced by the new cbmc) and without (produced by the old cbmc).  And the new cbmc does not break either of the old and new viewers.  The two versions of cbmc change their output enough that comparing across versions of cbmc is not practical.

The commits have been prepared carefully to make it easy to review this pull request commit by commit.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/cbmc_viewer/coveraget.py [hunk 1]
```diff
@@ -476,59 +476,56 @@ def update_coverage(coverage, path, func, line, status):
...
+    lines = set()
+    for group in string.split(','):
+        bounds = group.split('-')
+        if len(bounds) == 1:
+            lines.add(int(bounds[0]))
+            continue
+        if len(bounds) == 2:
+            lines.update(range(int(bounds[0]), int(bounds[1])+1))
+            continue
+        raise UserWarning(f"Unexpected encoding of line numbers: {string}")
```
Comment (tautschnig) [inline src/cbmc_viewer/coveraget.py:499]: Nit picking: why not write this as `if ... elif ... else`?

File: src/cbmc_viewer/coveraget.py [hunk 2]
```diff
@@ -476,59 +476,56 @@ def update_coverage(coverage, path, func, line, status):
...
-    lines = chunk[last_colon + 1:]
-    return [(filename, function, line) for line in parse_lines(lines)]
+# Extract the source lines contributing to a coverage goal's basic block
+#
+# Each coverage goal is a basic block in a goto program. Each
+# instruction in a basic block has a source location identifying the
+# source line that generated the instruction.
+
+def parse_lines(string):
+    "Extract line numbers from the string encoding of line numbers used in coverage output"
```
Comment (tautschnig) [inline src/cbmc_viewer/coveraget.py:484]: Why is this a string instead of a docstring?

File: src/cbmc_viewer/coveraget.py [hunk 3]
```diff
@@ -476,59 +476,56 @@ def update_coverage(coverage, path, func, line, status):
...
+        if len(bounds) == 2:
+            lines.update(range(int(bounds[0]), int(bounds[1])+1))
+            continue
+        raise UserWarning(f"Unexpected encoding of line numbers: {string}")
+    return sorted(lines)
 
-def parse_lines(lines):
-    """Parse the lines in a chunk of coverage"""
+def parse_description(description):
+    "Extract basic block source lines from a coverage goal's textual description"
```
Comment (tautschnig) [inline src/cbmc_viewer/coveraget.py:503]: Should this be a docstring instead?

File: src/cbmc_viewer/coveraget.py [hunk 4]
```diff
@@ -527,6 +531,51 @@ def parse_description(description):
         # ValueError in parse_lines(): invalid literal for int()
         raise UserWarning(f'Unexpected coverage goal description: "{description}"') from error
 
+def parse_basic_block_lines(basic_block_lines):
+    "Extract basic block source lines from a coverage goal's xml subelement"
```
Comment (tautschnig) [inline src/cbmc_viewer/coveraget.py:529]: Docstring?

File: src/cbmc_viewer/coveraget.py [hunk 5]
```diff
@@ -527,6 +531,51 @@ def parse_description(description):
...
+                 for fyle, func, lines in [(bbl.get("file"), bbl.get("function"), bbl.text)]
+                 for line in parse_lines(lines)]
+        assert all(fyle and func and line for fyle, func, line in lines)
+        return lines
+    except (ValueError, AssertionError) as error:
+        # ValueError in parse_lines(): invalid literal for int()
+        raise UserWarning(f'Unexpected coverage goal xml data: "{basic_block_lines}"') from error
+
+def parse_basicBlockLines(basicBlockLines): # pylint: disable=invalid-name
+    "Extract basic block source lines from a coverage goal's json data"
```
Comment (tautschnig) [inline src/cbmc_viewer/coveraget.py:551]: Docstring?

**Review by tautschnig — APPROVED** [review]

File: src/cbmc_viewer/coveraget.py (hunk 1 above)
Comment (markrtuttle) [reply src/cbmc_viewer/coveraget.py]: Done.

File: src/cbmc_viewer/coveraget.py (hunk 2 above)
Comment (markrtuttle) [reply src/cbmc_viewer/coveraget.py]: Done, although the difference between `"` and `"""` is string versus multi-line string and not string versus docstring.

File: src/cbmc_viewer/coveraget.py (hunk 3 above)
Comment (markrtuttle) [reply src/cbmc_viewer/coveraget.py]: Done

File: src/cbmc_viewer/coveraget.py (hunk 4 above)
Comment (markrtuttle) [reply src/cbmc_viewer/coveraget.py]: Done

File: src/cbmc_viewer/coveraget.py (hunk 5 above)
Comment (markrtuttle) [reply src/cbmc_viewer/coveraget.py]: Done

**Pull request merged**